lib/ash/changeset/changeset.ex

```diff
@@ -1094,50 +1094,58 @@ defmodule Ash.Changeset do
   end
 
   @doc false
+  # Returns either an appropriate expression for an atomic condition or a value
+  # indicated that the condition cannot be handled atomically.
+  #
+  # Validation logic matches on failure. So, for example, `present(:field)` is
+  # going to _match_ when `:field` is `nil`. However, when applying this logic
+  # to a `where` condition, the opposite is desired. The end result is kinda
+  # ugly because it can end up reading like "not is not equal to" but
+  # ultimately produces the correct results.
+  @spec atomic_condition([{module(), keyword()}], Ash.Changeset.t(), map()) ::
+          {:atomic, Ash.Expr.t() | boolean()} | {:not_atomic, String.t()}
   def atomic_condition(where, changeset, context) do
     Enum.reduce_while(where, {:atomic, true}, fn {module, validation_opts},
-                                                 {:atomic, condition} ->
+                                                 {:atomic, condition_expr} ->
       case module.atomic(
              changeset,
              validation_opts,
              struct(Ash.Resource.Validation.Context, context)
            ) do
         :ok ->
-          {:cont, {:atomic, condition}}
-
-        {:error, _} ->
-          {:cont, {:atomic, false}}
-
-        [{:atomic, _, expr, _as_error} | rest] ->
-          exprs = [expr | Enum.map(rest, &elem(&1, 2))]
-
-          new_expr =
-            Enum.reduce(exprs, condition, fn expr, condition ->
-              if condition == true do
-                expr
-              else
-                expr(^condition and ^expr)
-              end
-            end)
-
-          {:cont, {:atomic, new_expr}}
+          {:cont, {:atomic, condition_expr}}
 
         {:atomic, _, expr, _as_error} ->
-          new_expr =
-            if condition == true do
-              expr
-            else
-              expr(^condition and ^expr)
-            end
-
-          {:cont, {:atomic, new_expr}}
+          {:cont, {:atomic, atomic_condition_expr(condition_expr, expr)}}
 
-        {:not_atomic, reason} ->
-          {:halt, {:not_atomic, reason}}
+        {:error, _} ->
+          # Error from the validator, so the validations should just fail with
+          # a `false` expression.
+          {:halt, {:atomic, false}}
+
+        {:not_atomic, _reason} = not_atomic ->
+          {:halt, not_atomic}
+
+        atomic_conditions when is_list(atomic_conditions) ->
+          atomic_conditions
+          |> Stream.map(fn {:atomic, _, expr, _as_error} -> expr end)
+          |> Enum.reduce(condition_expr, &atomic_condition_expr(&2, &1))
```

Comment:
I'd probably just do the matching here, i.e `Enum.reduce(condition_expr, fn {:atomic, _, expr, _} -> ...` since streams have some overhead and aren't usually worth it on small collections.

Reply:
actually its such a tiny change I'll make it myself. No need to hold up this awesome PR ❤️

Reply:
yknow what, it doesn't matter. I'll leave it 😄

Reply:
I have had it in my head to always Stream and then finish with an Enum. I wonder where I picked that up. Should I avoid enums on small collections? (Any chance you have a source handy?)

Reply:
https://elixirforum.com/t/understanding-streams-and-enum/18574/2

Reply:
Ah-ha. From Valim himself:

> I like to say that streams are about using less memory at the cost of CPU.

I _think_ I got the idea from [this book](https://www.manning.com/books/elixir-in-action-third-edition) which was the first Elixir book I read (and I think it's great too). I'm going to have to dig it up and see how well my memory matches reality—at my age, that's becoming a more frequent issue.

Oh, and look at that... the author of said book [literally weights in](https://elixirforum.com/t/understanding-streams-and-enum/18574/5?u=binarypaladin) just after José. Welp, I need to revise my position and link this to my team, lol. Thanks.